File: src/main/java/ru/grabovsky/recordkeeping/models/entities/Record.java
package ru.grabovsky.recordkeeping.models.entities;

import lombok.*;
import org.hibernate.Hibernate;
import org.hibernate.annotations.CreationTimestamp;
import org.hibernate.annotations.Type;
import org.hibernate.annotations.TypeDef;
import org.hibernate.annotations.UpdateTimestamp;
import ru.grabovsky.recordkeeping.models.types.Direction;
import ru.grabovsky.recordkeeping.models.types.PostgreSQLEnumType;

import javax.persistence.*;
import java.time.Instant;
import java.time.LocalDate;
import java.util.List;
import java.util.Objects;

/**
 * Entity for record of incoming or outgoing documents
 *
 * @author GrabovskyAlexey
 */
@Entity
@Table(name = "records")
@TypeDef(name = "pg_enum", typeClass = PostgreSQLEnumType.class)
@Getter
@Setter
@ToString
public class Record {
    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    private Long id;

    @Type(type = "pg_enum")
    @Column(name = "direction", nullable = false)
    private Direction direction;

    @Column(name = "color", nullable = false)
    private String color;

    @Column(name = "mail_number", nullable = false)
    private String mailNumber;

    @Column(name = "reg_date", nullable = false)
    private LocalDate regDate;

    @Column(name = "title", nullable = false)
    private String title;

    @Column(name = "reply")
    private String reply;

    @Column(name = "reply_to")
    private String replyTo;

    @Column(name = "mail_date", nullable = false)
    private LocalDate mailDate;

    @Column(name = "description")

    private String description;
    @CreationTimestamp
    @Column(name = "created_at")
    private Instant createdAt;

    @UpdateTimestamp
    @Column(name = "updated_at")
    private Instant updatedAt;

    @ManyToOne
    @JoinColumn(name = "project_id")
    private Project project;

    @ManyToOne
    @JoinColumn(name = "employee_id")
    private Employee employee;

    @ManyToOne
    @JoinColumn(name = "contractor_id")
    private Contractor contractor;

    @ManyToOne
    @JoinColumn(name = "company_id", nullable = false)
    private Company company;

    @ManyToOne
    @JoinColumn(name = "author_id", nullable = false)
    private User author;

    @OneToMany(mappedBy = "record")
    @ToString.Exclude
    private List<File> files;

    public Record() {
        this.regDate = LocalDate.now();
        this.mailDate = LocalDate.now();
    }

    @Override
    public boolean equals(Object o) {
        if (this == o) return true;
        if (o == null || Hibernate.getClass(this) != Hibernate.getClass(o)) return false;
        Record record = (Record) o;
        return id != null && Objects.equals(id, record.id);
    }

    @Override
    public int hashCode() {
        return getClass().hashCode();
    }
}

File: core/src/main/kotlin/ru/grabovsky/recordkeeping/core/entity/db/Record.kt
package ru.grabovsky.recordkeeping.core.entity.db

import jakarta.persistence.*
import org.hibernate.annotations.CreationTimestamp
import org.hibernate.annotations.JdbcType
import org.hibernate.annotations.UpdateTimestamp
import org.hibernate.dialect.PostgreSQLEnumJdbcType
import ru.grabovsky.recordkeeping.api.types.Direction
import java.time.Instant
import java.time.LocalDate

/**
 * Entity for record of incoming or outgoing documents
 *
 * @author GrabovskyAlexey
 */
@Entity
@Table(name = "records")
class Record(
    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    override var id: Long? = null
): BaseEntity(id) {

    @Enumerated(EnumType.STRING)
    @JdbcType(PostgreSQLEnumJdbcType::class)
    @Column(columnDefinition = "direction", nullable = false)
    lateinit var direction: Direction

    @Column(name = "color", nullable = false)
    var color: String = "#FFFFFF"

    @Column(name = "mail_number", nullable = false)
    lateinit var mailNumber: String

    @Column(name = "reg_date", nullable = false)
    lateinit var regDate: LocalDate

    @Column(name = "subject", nullable = false)
    lateinit var subject: String

    @OneToMany(mappedBy = "replyTo", cascade = [CascadeType.ALL], orphanRemoval = true)
    @Column(name = "reply_id")
    val reply: Set<Record> = setOf()

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "reply_to_id")
    var replyTo: Record? = null

    @Column(name = "mail_date", nullable = false)
    lateinit var mailDate: LocalDate

    @Column(name = "description")
    var description: String? = null

    @CreationTimestamp
    @Column(name = "created_at")
    var createdAt: Instant? = null

    @UpdateTimestamp
    @Column(name = "updated_at")
    var updatedAt: Instant? = null

    @ManyToOne
    @JoinColumn(name = "project_id")
    var project: Project? = null

    @ManyToOne
    @JoinColumn(name = "employee_id")
    var employee: Employee? = null

    @ManyToOne
    @JoinColumn(name = "contractor_id")
    var contractor: Contractor? = null

    @ManyToOne
    @JoinColumn(name = "organization_id", nullable = false)
    lateinit var organization: Organization

    @ManyToOne
    @JoinColumn(name = "author_id", nullable = false)
    lateinit var author: User

    @OneToMany(mappedBy = "record")
    val files: List<File> = listOf()

    override fun toString(): String {
        return "Record(id=$id, direction=$direction, color='$color', mailNumber='$mailNumber', regDate=$regDate, subject='$subject', reply=$reply, replyTo=$replyTo, mailDate=$mailDate, description=$description, project=$project, employee=$employee, contractor=$contractor, organization=$organization, author=$author)"
    }
}